better-code/src/chapter-2-contracts.md

```diff
@@ -0,0 +1,1129 @@
+more efficient.
+
+## In pursuit of Correctness
+
+Fundamentally, contracts are about the first ingredient of Better
+Code: correctness. It's popular in some circles to scoff at the idea
+of correct software, on the grounds that bugs are inevitable, and
+because _for the most part_, we rely on incorrect software every day
+without disaster.  But there is a practical approach to correctness
+that improves both code quality and the programming experience,
+without demanding unrealistic perfection.
+
+For those who like to think formally, we can define a _semi-correct_
+program as one for which there exists a valid initial state and
+inputs, such that the program will generate a valid output. As an
+analogy, a broken clock is semi-correct, it is correct twice a
+day. This definition provides a partial ordering of correctness, where
+a program is _more correct_ than another if the initial states under
+which it is correct are a superset of the other.
+
+So while you may not reach a point where all bugs vanish, you will
+have well-justified confidence in the correctness of your own work
+while maintaining high productivity. Also, the discipline we'll
+describe here of **programming by contract** removes uncertainty and
+needless complexity from your code and from the process of coding.
+That's a big part of why we say it's practical.
+
+## Reasoning about correctness
+
+How can you know whether your program is correct?  Validating
+correctness seldom requires an elaborate formal proof—usually the sort
+of everyday thinking that we do while programming suffices.
+
+```swift
+var names = [ "Sean", "Laura", "Dave", "Crusty" ]
+names.sort()
+print(names[3])
+```
+
+In the example above, how do we know that last line is OK? Our thought
+process is something like this:
+
+- We started with values in indices zero through three of `names`.
+- sorting `names` rearranges elements without changing the set of valid indices.
+- Therefore, in the last line, `names` still has an element at index `3`.
+
+That's just an informal proof, and we use that kind of reasoning every
+time we write a line of code. So, regular programming is on the same
+continuum as formally proving correctness.  We're going to explore the
+power that's available by moving toward the formal end of that
+spectrum to see how it enables Better Code.
+
+## Local Reasoning
+
+The kind of everyday thinking demonstrated above is only practical if
+we can reason locally about code.  The clearest definition of Local
+Reasoning we've found is [this
+one](https://medium.com/@nathangitter/local-reasoning-in-swift-6782e459d)
+from Nathan Gitter: “Local reasoning is the idea that the reader can
+make sense of the code directly in front of them, without going on a
+journey discovering how the code works.”
+
+So in our example, what we know about `sort` allow us to reason about
+its use without looking at its implementation.  In fact, local
+reasoning is so fundamental that most of our programming best
```

Comment:
```suggestion
reasoning is so fundamental that most programming best
```

With, "our", did you mean STLab's? Or the industry's?

Reply:
I meant the industry's.  I think I might drop “programming” as well.

```suggestion
reasoning is so fundamental that most accepted best
```